aiida/tools/importexport/dbimport/readers.py

```diff
@@ -0,0 +1,328 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Archive reader classes."""
+from abc import ABC, abstractmethod
+import dataclasses
+import json
+import os
+from typing import Any, Dict, Iterable, List, Optional, Set, Tuple
+import zipfile
+import tarfile
+
+from distutils.version import StrictVersion
+from wrapt import decorator
+
+from aiida.common.exceptions import InvalidOperation
+from aiida.common.folders import Folder, SandboxFolder
+from aiida.tools.importexport.common.config import EXPORT_VERSION, NODES_EXPORT_SUBFOLDER
+from aiida.tools.importexport.common.exceptions import (
+    CorruptArchive, ImportValidationError, IncompatibleArchiveVersionError
+)
+from aiida.tools.importexport.common.utils import export_shard_uuid
+
+
+def detect_file_type(in_path):
+    """For back-compatibility."""
+    if os.path.isdir(in_path):
+        return 'folder'
+    if tarfile.is_tarfile(in_path):
+        return 'tar'
+    if zipfile.is_zipfile(in_path):
+        return 'zip'
+    raise ImportValidationError(
+        'Unable to detect the input file format, it is neither a '
+        'tar file, nor a (possibly compressed) zip file.'
+    )
+
+
+@dataclasses.dataclass
+class ExportMetadata:
+    """Class for storing metadata about an export."""
+    graph_traversal_rules: Optional[Dict[str, bool]]
+    # Entity type -> UUID list
+    entities_starting_set: Optional[Dict[str, Set[str]]]
+    include_comments: Optional[bool]
+    include_logs: Optional[bool]
+    # Entity type -> database ID key
+    unique_identifiers: Dict[str, str] = dataclasses.field(repr=False)
+    # Entity type -> database key -> meta parameters
+    all_fields_info: Dict[str, Dict[str, Dict[str, str]]] = dataclasses.field(repr=False)
+    aiida_version: str
+    export_version: str
+
+
+@decorator
+def ensure_within_context(wrapped, instance, args, kwargs):
```

Comment:
this seems like a widely applicable decorator - can it be made available more centrally?

Reply:
yeh possibly, although obviously it is only applicable to classes that set/unset this `_in_context` attribute in their `__enter__`/`__exit__` methods

Reply:
I see. I believe something similar was already added here in a different context:
https://github.com/aiidateam/aiida-core/blob/02248cf3686a0ab89faf1625e0da24d9e33d8cde/aiida/orm/nodes/node.py#L39

Reply:
yep I recall conversation about that ta 👍 
I guess it is potatas/potatas on whether to add this as deorator or a call to a method (maybe a call to a method is better though, because it does not have any limitations e.g. for `@property`)

I actually copied the decorator from `aiida/tools/importexport/common/archive.py:Archive` (note this class also needs to be deprecated, to replaced by `ReaderJsonZip` etc). There it was used as a method on the class, but this did not work with class inheritance

Reply:
I'm not sure how we want to go about the deprecation of `aiida/tools/importexport/common/archive.py` and `aiida/tools/importexport/common/progress_bar.py`. I guess they need to be left in for aiida v1.5.0. possibly you can put a deprecation warning in the module itself, i.e. to trigger when the module is imported, or you put them in all the contained functions/classes

Reply:
> I'm not sure how we want to go about the deprecation of `aiida/tools/importexport/common/archive.py` and `aiida/tools/importexport/common/progress_bar.py`.
> I guess they need to be left in for aiida v1.5.0.
> Possibly you can put a deprecation warning in the module itself, i.e. to trigger when the module is imported, or you put them in all the contained functions/classes

In terms of import depth, I think both modules are far enough down not to be considered part of the official public API 
(I was never quite sure of the exact rules; I've just opened a PR to clarify them https://github.com/aiidateam/aiida-core/pull/4511).

In any case, I care more about *actual* uses in the wild, and I highly doubt that the progress bar was used by any plugin (I've never seen/heard of this). I.e. I'd be fine with removing `aiida/tools/importexport/common/progress_bar.py` in 1.5.0

For the `export_tar`,... functions in `archive.py` I know that they actually have been used in the wild, so here we need to provide a deprecation path. A deprecation warning at the module level can work as long as it is possible to provide a warning message that clearly explains what to do for any of the deprecated functionality (otherwise it's better to keep one per function).

For the `Archive` class, I have personally never seen it used and I'm having a tough time imagining anyone using it outside aiida-core, although it is possible. A [quick github search](https://github.com/search?p=2&q=aiida+Archive+unpack&type=Code) did not turn up anything.
Here I would leave it up to you if you want to leave it in.

Reply:
I've changed the decorator to be a method